Letters use new address mg

markgreenburg opened the pull request:
- Switch to using address returned from API in the UI after a successful address update (involved some transforming of the response back into the shape required by the front end)
- Log any mismatches between PUT and response addresses to Sentry
- Update related tests

Also successfully tested all update permutations manually against CI (M -> D; M -> I; M -> M; etc...)

markgreenburg committed:
update saveAddress to compare and use response, update tests

markgreenburg committed:
Add function to remove empty values from address

markgreenburg committed:
Put empty value function in helpers

markgreenburg committed:
Update tests to use real-life keys

markgreenburg committed:
Make helper more obvious

markgreenburg committed:
Use slow timeout

markgreenburg committed:
WIP commit

markgreenburg committed:
Remove old code

markgreenburg committed:
Update test data

markgreenburg committed:
Add tests for helpers

markgreenburg committed:
Un-restrict unit tests

cvalarida commented on src/js/letters/actions/letters.js:
```diff
@@ -236,13 +228,19 @@ export function saveAddress(address) {
   };
   window.dataLayer.push({ event: 'letter-update-address-submit' });
   return (dispatch) => {
-    // TODO: Show a spinner or some kind of indication we're waiting on this to return
```
👍

cvalarida commented on src/js/letters/actions/letters.js:
```diff
@@ -236,13 +228,19 @@ export function saveAddress(address) {
+        const responseAddress = militaryToGeneric(response.data.attributes.address);
+        if (!isEqual(stripEmpties(address), stripEmpties(responseAddress))) {
+          const mismatchError = new Error('letters-address-update addresses don\'t match');
+          Raven.captureException(mismatchError, { submitted: address, returned: responseAddress });
```
Let's make sure we can log this before we merge it in. Not sure if it counts as PII.

markgreenburg replied:
Oh, that's a great call...I'll check into it

markgreenburg replied:
Checked with @kreek. Seems as though you could reverse engineer PII from the address itself so changed the code to ensure we don't log any of the actual address information itself.

cvalarida commented on src/js/letters/utils/helpers.jsx:
```diff
@@ -339,3 +339,37 @@ export function resetDisallowedAddressFields(address) {
+ * @returns a cloned object with no empty properties
+ */
+export const stripEmpties = (input) => {
+  const newObject = { ...input };
```
I just [confirmed on jsbin](http://jsbin.com/qifequzure/edit?js,console) that passing variables to a function does a shallow clone, so this line doesn't actually do anything. If you're looking to not modify `input`, you'll want to `_.deepClone()`.

cvalarida replied:
Er, I stand corrected. Passing happens by reference, so cloning is necessary.

This is still a shallow clone, but that's probably fine since the address passed as `input` isn't going to have any sub-objects.

markgreenburg replied:
Yeah, I still don't think it's super great that this fact isn't actually enforced by the code, but don't think we need to do anything about it right now.

cvalarida commented on src/js/letters/utils/helpers.jsx:
```diff
@@ -339,3 +339,37 @@ export function resetDisallowedAddressFields(address) {
+ */
+export const stripEmpties = (input) => {
+  const newObject = { ...input };
+  const deleteProperty = (key) => (delete newObject[key]);
```
Out of idle curiosity, why split this out here instead of defining the function directly in the `.forEach()` call?

cvalarida commented on test/e2e/letters-helpers.js:
```diff
@@ -74,19 +74,47 @@ const address = {
+//   type: 'DOMESTIC',
+//   zipCode: '60602',
+//   zipSuffix: ''
+// };
```
Can probably get rid of this.

cvalarida commented on test/letters/actions/letters.unit.spec.js:
```diff
@@ -60,8 +60,21 @@ const getState = () => ({});
   const frontEndAddress = {
     type: ADDRESS_TYPES.military,
-    city: 'apo',
-    state: 'secret'
```
Ahh, should have been `state: 'Confusion'`! 😜 
Too bad!

cvalarida commented on src/js/letters/utils/helpers.jsx:
```diff
@@ -339,3 +339,37 @@ export function resetDisallowedAddressFields(address) {
+  if (address.type !== ADDRESS_TYPES.military) {
+    return { ...address };
+  }
+  const genericAddress = { ...address };
```
Same with these; you might want to `_.cloneDeep()`.

Not that it's a huge deal. :man_shrugging:

cvalarida commented on test/letters/utils/helpers.unit.spec.jsx:
```diff
@@ -135,5 +137,90 @@ describe('Letters helpers: ', () => {
+      const militaryTest = militaryToGeneric(militaryAddress);
+      const nonMilitaryTest = militaryToGeneric(domesticAddress);
+      expect(militaryTest).to.not.equal(militaryAddress);
+      expect(nonMilitaryTest).to.not.equal(domesticAddress);
```
See, this is tricky because the "top-level" objects _are_ different refs, but the "sub" objects will have the same refs.

For the address objects, though, it's probably totally fine because we won't have any nested objects. Suppose this is more academic then anything, really. 🤔

markgreenburg replied:
Yeah, so these are just shallow objects with string values. But, that's just an implicit invariant and not actually specced out anywhere in the codebase, which is probably not great.

cvalarida reviewed (commented):
Looks good. Despite my repeated comments about not actually cloning the address, I think it's golden. The only reason I'm not going to approve this right now is because I want to make sure we can actually log the user's address first.

markgreenburg committed:
Do not log address data in Sentry error log

markgreenburg committed:
Remove old code that was commented out anyway

markgreenburg commented:
@cvalarida Addressed some of your comments and cleared the address logging with @kreek. And by 'cleared', I mean 'I changed the code by not logging the actual addresses to Sentry' 😉 
Would appreciate another run-through when you're feeling up to it so we can get this sucker merged. Thanks!

cvalarida reviewed (approved):
:shipit:
